src/cpp/fastdds/publisher/DataWriterImpl.cpp

```diff
@@ -226,29 +226,100 @@ ReturnCode_t DataWriterImpl::write(
+        logError(PUBLISHER, "Data pointer not valid");
+        return c_InstanceHandle_Unknown;
     }
-    logInfo(DATA_WRITER, "Disposing of data");
-    WriteParams wparams;
-    if (create_new_change_with_params(NOT_ALIVE_DISPOSED, data, wparams, handle))
+
+    if (!type_->m_isGetKeyDefined)
     {
-        return ReturnCode_t::RETCODE_OK;
+        logError(PUBLISHER, "Topic is NO_KEY, operation not permitted");
+        return c_InstanceHandle_Unknown;
     }
-    return ReturnCode_t::RETCODE_ERROR;
+
+    InstanceHandle_t instance_handle = c_InstanceHandle_Unknown;
+    bool is_key_protected = false;
+#if HAVE_SECURITY
+    is_key_protected = writer_->getAttributes().security_attributes().is_key_protected;
+#endif
+    type_->getKey(key, &instance_handle, is_key_protected);
+
+    // Block lowlevel writer
+    auto max_blocking_time = std::chrono::steady_clock::now() +
+            std::chrono::microseconds(::TimeConv::Time_t2MicroSecondsInt64(qos_.reliability().max_blocking_time));
+
+#if HAVE_STRICT_REALTIME
+    std::unique_lock<RecursiveTimedMutex> lock(writer_->getMutex(), std::defer_lock);
+    if (lock.try_lock_until(max_blocking_time))
+#else
+    std::unique_lock<RecursiveTimedMutex> lock(writer_->getMutex());
+#endif
+    {
+        if (history_.register_instance(instance_handle, lock, max_blocking_time))
+        {
+            return instance_handle;
+        }
+    }
+
+    return c_InstanceHandle_Unknown;
 }
 
-bool DataWriterImpl::dispose(
-        void* data)
+ReturnCode_t DataWriterImpl::unregister_instance(
+        void* instance,
+        const InstanceHandle_t& handle,
+        bool dispose)
 {
-    logInfo(DATA_WRITER, "Disposing of data");
-    WriteParams wparams;
-    return create_new_change_with_params(NOT_ALIVE_DISPOSED, data, wparams);
+    /// Preconditions
+    if (instance == nullptr)
+    {
+        logError(PUBLISHER, "Data pointer not valid");
+        return ReturnCode_t::RETCODE_BAD_PARAMETER;
+    }
+
+    if (!type_->m_isGetKeyDefined)
+    {
+        logError(PUBLISHER, "Topic is NO_KEY, operation not permitted");
+        return ReturnCode_t::RETCODE_PRECONDITION_NOT_MET;
+    }
+
+    ReturnCode_t returned_value = ReturnCode_t::RETCODE_ERROR;
+    InstanceHandle_t ih = handle;
+
+#if !defined(NDEBUG)
+    if (c_InstanceHandle_Unknown == ih)
+#endif
+    {
+        bool is_key_protected = false;
+#if HAVE_SECURITY
+        is_key_protected = writer_->getAttributes().security_attributes().is_key_protected;
+#endif
+        type_->getKey(instance, &ih, is_key_protected);
+    }
+
+#if !defined(NDEBUG)
+    if (c_InstanceHandle_Unknown != handle && ih != handle)
+    {
+        logError(PUBLISHER, "handle differs from data's key.");
+        return ReturnCode_t::RETCODE_BAD_PARAMETER;
+    }
+#endif
+
+    if (history_.key_is_registered(ih))
+    {
+        WriteParams wparams;
+        ChangeKind_t change_kind = dispose ? NOT_ALIVE_DISPOSED : NOT_ALIVE_UNREGISTERED;
```

Comment:
I think we should use `NOT_ALIVE_DISPOSED_UNREGISTERED`, as it is the value being compared inside `onWriterChangeReceivedByAll`
```suggestion
        ChangeKind_t change_kind = dispose ? NOT_ALIVE_DISPOSED_UNREGISTERED : NOT_ALIVE_UNREGISTERED;
```

Reply:
The same function is used to unregister an instance (sending a NO_ALIVE_UNREGISTERED), or disposing an instance (sending a NO_ALIVE_DISPOSED). I've added a new possibility. If the instance is being unregistered and `writer_data_lifecycle.autodispose_unregistered_instances` is `true`, then NO_ALIVE_DISPOSED_REGISTERED is sent.